log/src/lib.rs

```diff
@@ -0,0 +1,167 @@
+use lv2_core::feature::{Feature, ThreadingClass};
+//use std::ffi::CString;
+use std::os::raw::*; //get all common c_type
+use urid::*;
+
+pub struct EntryClass;
+unsafe impl UriBound for EntryClass {
+    const URI: &'static [u8] = lv2_sys::LV2_LOG__Entry;
+}
+
+/// UriBound for an error message.
+pub struct ErrorClass;
+unsafe impl UriBound for ErrorClass {
+    const URI: &'static [u8] = lv2_sys::LV2_LOG__Error;
+}
+
+/// UriBound for an informative message.
+pub struct NoteClass;
+unsafe impl UriBound for NoteClass {
+    const URI: &'static [u8] = lv2_sys::LV2_LOG__Note;
+}
+
+/// UriBound for a debuging message.
+pub struct TraceClass;
+unsafe impl UriBound for TraceClass {
+    const URI: &'static [u8] = lv2_sys::LV2_LOG__Trace;
+}
+
+/// Uribound for an error message.
+pub struct WarningClass;
+unsafe impl UriBound for WarningClass {
+    const URI: &'static [u8] = lv2_sys::LV2_LOG__Warning;
+}
+
+/// Marker for URID representing the nature of a log message.
+// Note : it's may be better to have a URID trait to define a common interface
+pub unsafe trait EntryType
+where
+    Self: Copy,
+{
+}
+
+unsafe impl EntryType for URID<ErrorClass> {}
+unsafe impl EntryType for URID<NoteClass> {}
+unsafe impl EntryType for URID<TraceClass> {}
+unsafe impl EntryType for URID<WarningClass> {}
+
+/// Errors potentially generated by [`Log`](struct.Log.html) methods.
+#[derive(PartialEq, Eq, Clone, Copy, Debug)]
+pub enum LogError {
+    /// An error occured when sending message to the host
+    PrintError,
+    /// The provided string don't have a `'\0'` to indicate it's end
+    NoNullTerminator,
+    /// No callback was provided by the host
+    ///
+    /// This can only happen with a faulty host
+    NoCallback,
+}
+
+/// The Log feature.
+#[repr(transparent)]
+pub struct Log<'a> {
+    internal: &'a lv2_sys::LV2_Log_Log,
+}
+
+unsafe impl<'a> UriBound for Log<'a> {
+    const URI: &'static [u8] = lv2_sys::LV2_LOG__log;
+}
+
+unsafe impl<'a> Feature for Log<'a> {
+    // Note: this feature can be used in any threading class but:
+    // * i have a doubt about it's thread safety, can we assume the host provide this thread safety
+    // since this feature can be used anywhere ?.
+    // * i shouldn't be used in context where rt is a concern, that mean is audiothreadclass in
+    // practice, but it's acceptable to use it for debugging purpose
+    // So, at this time, i just giving access to it instanciation class
+    unsafe fn from_feature_ptr(feature: *const c_void, class: ThreadingClass) -> Option<Self> {
+        match class {
```

Comment:
To answer your question here: this feature can be used in any threading class, so the `class` parameter can just be ignored. :slightly_smiling_face:

Reply:
yes, the features can be used in any threading class, but only "trace" should be used in audio thread, and it's only intended for debug reason. At the moment, to handle this, i only see optional feature to disable/enable this feature in audio context

Reply:
lol, mistake, i confused rt-context with audio thread context. a Rt-safety paragraph should be added somewhere